laplace/curvature/curvature.py

```diff
@@ -188,38 +229,39 @@ class GGNInterface(CurvatureInterface):
         to apply the Laplace approximation over
     stochastic : bool, default=False
         Fisher if stochastic else GGN
+    num_samples: int, default=100
+        Number of samples used to approximate the stochastic Fisher
     """
-    def __init__(self, model, likelihood, last_layer=False, subnetwork_indices=None, stochastic=False):
+    def __init__(self, model, likelihood, last_layer=False, subnetwork_indices=None, stochastic=False, num_samples=1):
         self.stochastic = stochastic
+        self.num_samples = num_samples
         super().__init__(model, likelihood, last_layer, subnetwork_indices)
 
-    def _get_full_ggn(self, Js, f, y):
-        """Compute full GGN from Jacobians.
+    def _sample_H_lik(self, f):
+        H_lik = 0
 
-        Parameters
-        ----------
-        Js : torch.Tensor
-            Jacobians `(batch, parameters, outputs)`
-        f : torch.Tensor
-            functions `(batch, outputs)`
-        y : torch.Tensor
-            labels compatible with loss
+        for _ in range(self.num_samples):
+            if self.likelihood == 'regression':
+                y_sample = f + torch.randn(f.shape, device=f.device)  # N(y | f, 1)
+                grad_sample = f - y_sample  # functional MSE grad
+            else:  # classification with softmax
+                y_sample = torch.distributions.Multinomial(logits=f).sample()
+                # First functional derivative of the loglik is p - y
+                p = torch.softmax(f, dim=-1)
+                grad_sample = p - y_sample
 
-        Returns
-        -------
-        loss : torch.Tensor
-        H_ggn : torch.Tensor
-            full GGN approximation `(parameters, parameters)`
-        """
-        loss = self.factor * self.lossfunc(f, y)
+            H_lik += 1/self.num_samples * torch.einsum('bc,bk->bck', grad_sample, grad_sample)
+
+        return H_lik
+
+    def _get_exact_H_lik(self, f):
```

Comment:
Can we leave this is `_get_full_ggn` and add the function `_get_full_fisher` instead of these renamings? I `exact_H_lik` is inappropriate because it is not the true Hessian and would in principle have to be `log_lik`. `sample_H_lik` also sounds a bit different from what it actually is and just using `fisher` seems more appropriate to me.

Reply:
I renamed them to better describe the functions. I think it's best to have them instead since there are many code overlaps in the `_get_full_ggn` and the proposed `_get_full_fisher` (also overlap with `diag_ggn` and `diag_fisher`).